File: shared/session-names.ts
/**
 * Canonical coding-agent session names — shared wire contract
 * (unified-agent-names plan, Task 1).
 *
 * One saved name per scoped Claude/Codex/OpenCode session (terminal modes
 * `claude`/`codex`/`opencode` and fresh types `freshclaude`/`freshcodex`/
 * `freshopencode`). The Rust server's `SessionNames` store is the authority;
 * these Zod schemas are the language-neutral wire shapes mirrored by
 * `crates/freshell-protocol/src/session_names.rs`.
 *
 * Wire rules:
 * - `NameRevision`-typed counters are JS-safe integers (`0 ..= 2^53-1`).
 * - `SessionNameRef` keys encode as a JSON discriminated TUPLE
 *   (`sessionNameRefKey`) — never colon-split opaque provider session IDs.
 * - Kilroy is out of scope: it shares the Claude runtime but keeps its
 *   existing naming behavior.
 */
import { z } from 'zod'

export const NamedProviderSchema = z.enum(['claude', 'codex', 'opencode'])
export type NamedProvider = z.infer<typeof NamedProviderSchema>

/** The six scoped coding-agent modes: three terminal CLIs + three fresh types. */
export const UnifiedAgentModeSchema = z.enum([
  'claude',
  'codex',
  'opencode',
  'freshclaude',
  'freshcodex',
  'freshopencode',
])
export type UnifiedAgentMode = z.infer<typeof UnifiedAgentModeSchema>

/** Who asked for a name: a human (`user`) or automation (`automatic`). */
export const NameIntentSchema = z.enum(['user', 'automatic'])
export type NameIntent = z.infer<typeof NameIntentSchema>

/**
 * Provenance of an accepted name. `legacy_protected` is assigned ONLY by the
 * Task 7 migration: its historic origin is unknown (never claimed as human).
 */
export const NameSourceSchema = z.enum([
  'manual',
  'legacy_protected',
  'freshell_ai',
  'provider_ai',
  'first_message',
  'directory',
])
export type NameSource = z.infer<typeof NameSourceSchema>

export const LEGACY_ORIGIN_UNKNOWN = 'unknown' as const

/** Monotonically allocated, JS-safe revision counter (0 ..= 2^53-1). */
export const NameRevisionSchema = z.number().int().min(0).max(Number.MAX_SAFE_INTEGER)
export type NameRevision = z.infer<typeof NameRevisionSchema>

/** Hard accepted-name cap in Unicode scalar values — mirrors the Rust store. */
export const MAX_NAME_SCALARS = 200

/** Unicode control characters (category Cc) — exactly Rust's `char::is_control`. */
const CONTROL_CHARACTERS = /[\u0000-\u001f\u007f-\u009f]/

/**
 * Accepted-name text, mirroring the Rust store's `validate_name` exactly so a
 * client-side-valid name can never fail server-side validation: non-empty
 * after trimming, at most 200 Unicode scalar values, and no control
 * characters anywhere (display safety — deliberately stricter than the
 * plan's control-ONLY rejection, kept on both sides of the wire).
 */
export const SessionNameTextSchema = z.string().superRefine((name, ctx) => {
  if (name.trim().length === 0) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: 'a session name cannot be empty or whitespace-only',
      path: [],
    })
    return
  }
  if ([...name].length > MAX_NAME_SCALARS) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: `a session name cannot exceed ${MAX_NAME_SCALARS} characters`,
      path: [],
    })
    return
  }
  if (CONTROL_CHARACTERS.test(name)) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: 'a session name cannot contain control characters',
      path: [],
    })
  }
})

/**
 * Naming target: a pre-durable pending handle (minted per logical
 * conversation before provider identity exists) or the durable
 * provider/session identity. The session variant reuses the existing
 * structured provider/session identity — no new restore identity.
 */
export const SessionNameRefSchema = z.discriminatedUnion('kind', [
  z.object({ kind: z.literal('pending'), id: z.string().min(1) }),
  z.object({ kind: z.literal('session'), provider: NamedProviderSchema, sessionId: z.string().min(1) }),
])
export type SessionNameRef = z.infer<typeof SessionNameRefSchema>

/**
 * Which session names a tab: `{kind:'session',paneId}` follows that pane's
 * canonical session name; `{kind:'legacy'}` keeps the existing non-agent
 * derivation. A tab stores the relationship, never a second name.
 */
export const TabNameSourceSchema = z.discriminatedUnion('kind', [
  z.object({ kind: z.literal('session'), paneId: z.string().min(1) }),
  z.object({ kind: z.literal('legacy') }),
])
export type TabNameSource = z.infer<typeof TabNameSourceSchema>

/**
 * One accepted name record. `manualRevision`/`renamedAt` are assigned only by
 * an actual explicit user rename; `legacyOrigin` is required (and only
 * legal) for `legacy_protected`.
 */
export const SessionNameRecordSchema = z
  .object({
    ref: SessionNameRefSchema,
    name: SessionNameTextSchema,
    source: NameSourceSchema,
    revision: NameRevisionSchema,
    manualRevision: NameRevisionSchema.optional(),
    renamedAt: z.number().int().nonnegative().optional(),
    legacyOrigin: z.literal(LEGACY_ORIGIN_UNKNOWN).optional(),
  })
  .superRefine((record, ctx) => {
    // Only migration assigns legacy_protected, and it must never claim human
    // origin: legacyOrigin is required there and illegal everywhere else.
    if (record.source === 'legacy_protected' && record.legacyOrigin !== LEGACY_ORIGIN_UNKNOWN) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'legacyOrigin is required for legacy_protected records',
        path: ['legacyOrigin'],
      })
    }
    if (record.source !== 'legacy_protected' && record.legacyOrigin !== undefined) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'legacyOrigin is only legal on legacy_protected records',
        path: ['legacyOrigin'],
      })
    }
  })
export type SessionNameRecord = z.infer<typeof SessionNameRecordSchema>

/** Pending→durable redirect, so late operations on a bound handle resolve. */
export const SessionNameRedirectSchema = z.object({
  from: SessionNameRefSchema,
  to: SessionNameRefSchema,
  revision: NameRevisionSchema,
})
export type SessionNameRedirect = z.infer<typeof SessionNameRedirectSchema>

/** Native writeback status (Task 3): display/status data, never a name authority. */
export const NativeSyncStatusSchema = z.enum(['pending', 'synced', 'unsynced', 'unsupported'])
export type NativeSyncStatus = z.infer<typeof NativeSyncStatusSchema>

/**
 * The native writeback status projected alongside a record (Task 3).
 * `synced` requires a current-revision readback of the exact desired name at
 * the attempted location; `unsynced` retains the uncertainty/exhaustion
 * reason; `unsupported` is a diagnosed capability failure. Status-only
 * updates fold on the client by `documentGeneration`, not name revision.
 */
export const NativeSyncSchema = z.object({
  status: NativeSyncStatusSchema,
  desiredRevision: NameRevisionSchema,
  locationRevision: NameRevisionSchema,
  observedCurrent: z.boolean().optional(),
  reason: z.string().optional(),
})
export type NativeSync = z.infer<typeof NativeSyncSchema>

/** The common response/broadcast shape for every naming operation. */
export const SessionNameUpdateSchema = z.object({
  record: SessionNameRecordSchema,
  documentGeneration: NameRevisionSchema,
  redirects: z.array(SessionNameRedirectSchema),
  changed: z.boolean(),
  nativeSync: NativeSyncSchema.optional(),
})
export type SessionNameUpdate = z.infer<typeof SessionNameUpdateSchema>

/** Rename request (HTTP PATCH body / MCP-bridged rename). */
export const RenameSessionNameRequestSchema = z.object({
  target: SessionNameRefSchema,
  name: SessionNameTextSchema,
  nameIntent: NameIntentSchema.optional(),
  ifRevision: NameRevisionSchema.optional(),
})
export type RenameSessionNameRequest = z.infer<typeof RenameSessionNameRequestSchema>

const TERMINAL_UNIFIED_MODES = new Set(['claude', 'codex', 'opencode'])
const FRESH_UNIFIED_SESSION_TYPES = new Set(['freshclaude', 'freshcodex', 'freshopencode'])

/**
 * Scope predicate: is this pane content one of the six unified naming modes?
 * Terminal scope is `mode in {claude,codex,opencode}`; fresh scope is
 * `sessionType in {freshclaude,freshcodex,freshopencode}` (fresh types also
 * match when passed as the mode, since fresh panes identify by session type).
 * Kilroy is explicitly out of scope even though it shares the Claude
 * runtime — a kilroy sessionType is never scoped, regardless of mode.
 */
export function isUnifiedAgentMode(mode: string | undefined, sessionType?: string): boolean {
  // Kilroy shares the Claude runtime but keeps its existing naming behavior:
  // a kilroy sessionType is never scoped, regardless of the mode it rides on.
  if (sessionType === 'kilroy') return false
  // Fresh panes identify by sessionType (fresh types also match when a caller
  // passes one as the mode).
  if (sessionType !== undefined && FRESH_UNIFIED_SESSION_TYPES.has(sessionType)) return true
  return mode !== undefined && (TERMINAL_UNIFIED_MODES.has(mode) || FRESH_UNIFIED_SESSION_TYPES.has(mode))
}

/**
 * Stable key for a naming ref: the JSON encoding of a discriminated TUPLE
 * (["pending",id] / ["session",provider,sessionId]) — never colon-splitting
 * opaque provider session IDs.
 */
export function sessionNameRefKey(ref: SessionNameRef): string {
  return ref.kind === 'pending'
    ? JSON.stringify(['pending', ref.id])
    : JSON.stringify(['session', ref.provider, ref.sessionId])
}

File: crates/freshell-protocol/src/session_names.rs
//! Canonical coding-agent session-name wire types (unified-agent-names plan,
//! Task 1), mirroring `shared/session-names.ts` exactly.
//!
//! * Wire enum strings are identical across TS/Rust.
//! * Rust fields are snake_case with camelCase Serde serialization.
//! * `NameRevision` is `u64` with a JS-safe integer ceiling
//!   ([`MAX_NAME_REVISION`]) — every revision/generation counter that crosses
//!   the wire to a JS client stays inside `Number.MAX_SAFE_INTEGER`.
//! * Serialization only, no logic: acceptance/rank rules live in the
//!   `freshell-server` store, not here.

use serde::{Deserialize, Serialize};

use crate::SessionLocator;

/// Monotonically allocated name revision — JS-safe on the wire.
pub type NameRevision = u64;

/// `Number.MAX_SAFE_INTEGER` (`2^53 - 1`): the hard ceiling for every
/// revision / document-generation counter serialized to a JS client.
pub const MAX_NAME_REVISION: NameRevision = 9_007_199_254_740_991;

/// The three scoped coding-agent providers.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum NamedProvider {
    Claude,
    Codex,
    Opencode,
}

impl NamedProvider {
    /// The exact wire string (`claude | codex | opencode`).
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Claude => "claude",
            Self::Codex => "codex",
            Self::Opencode => "opencode",
        }
    }

    /// Human-facing fallback label used when no directory basename exists.
    pub fn display_label(&self) -> &'static str {
        match self {
            Self::Claude => "Claude",
            Self::Codex => "Codex",
            Self::Opencode => "OpenCode",
        }
    }
}

/// Who asked for a name change: a human (`user`) or automation
/// (`automatic`). Default omitted intents to `automatic` on every scoped
/// create/rename route — an agent suggestion must never acquire the
/// permanence of a user's explicit rename.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum NameIntent {
    User,
    Automatic,
}

/// Provenance of an accepted name. Rank order for new operations is
/// `manual > legacy_protected > freshell_ai > provider_ai > first_message >
/// directory`; `legacy_protected` is assigned ONLY by the Task 7 migration
/// and never claims human origin.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum NameSource {
    Manual,
    LegacyProtected,
    FreshellAi,
    ProviderAi,
    FirstMessage,
    Directory,
}

/// The single legal `legacyOrigin` value: the historic origin of a
/// migration-protected label is unknown.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum LegacyOrigin {
    Unknown,
}

/// Naming target: a pre-durable pending handle (`pending`), or the durable
/// provider/session identity (`session`). The session variant reuses the
/// existing structured provider/session identity — no new restore identity.
/// Keys encode as a JSON discriminated tuple (see the store's `name_ref_key`)
/// and never colon-split opaque provider session IDs.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum SessionNameRef {
    #[serde(rename_all = "camelCase")]
    Pending { id: String },
    #[serde(rename_all = "camelCase")]
    Session {
        provider: NamedProvider,
        session_id: String,
    },
}

/// Which session names a tab: the source pane's canonical session name
/// (`session`), or the existing non-agent derivation (`legacy`). A tab stores
/// the relationship, never a second name.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum TabNameSource {
    #[serde(rename_all = "camelCase")]
    Session {
        pane_id: String,
    },
    Legacy,
}

/// One accepted name record. `manual_revision`/`renamed_at` are assigned only
/// by an actual explicit user rename; `legacy_origin` is required for (and
/// only legal on) `legacy_protected` records.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionNameRecord {
    #[serde(rename = "ref")]
    pub name_ref: SessionNameRef,
    pub name: String,
    pub source: NameSource,
    pub revision: NameRevision,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub manual_revision: Option<NameRevision>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub renamed_at: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub legacy_origin: Option<LegacyOrigin>,
}

/// Pending→durable redirect: late operations addressing a bound pending
/// handle resolve to the durable session record.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionNameRedirect {
    pub from: SessionNameRef,
    pub to: SessionNameRef,
    pub revision: NameRevision,
}

/// Native synchronization status for a canonical name's provider writeback
/// (unified-agent-names plan, Task 3): `pending` (a bounded native series is
/// armed/underway), `synced` (a current-revision readback confirmed the
/// provider holds the exact desired name/location), `unsynced` (the finite
/// allowance is exhausted or an outcome is ambiguous/divergent), and
/// `unsupported` (a diagnosed capability failure — unsupported/archived/
/// ephemeral/missing). Display/status data only: never another name
/// authority, and never a public native path.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum NativeSyncStatus {
    Pending,
    Synced,
    Unsynced,
    Unsupported,
}

/// The native writeback status projected alongside a record in
/// `SessionNameUpdate`/`SessionNameUpdated`. `desired_revision` is the
/// canonical record revision the native series is projecting; the provider is
/// `synced` only with that exact revision. `location_revision` is the routing
/// evidence revision the series attempted (0 when no verified location
/// exists). `observed_current` records that a readback observed the desired
/// name while an ambiguous outcome kept the status `unsynced`. `reason` is a
/// stable machine-facing diagnostic (never a prompt or native payload).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeSyncProjection {
    pub status: NativeSyncStatus,
    pub desired_revision: NameRevision,
    pub location_revision: NameRevision,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub observed_current: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

/// The common response/broadcast shape for every naming operation. `changed`
/// is true only when the accepted record (text/source/revision) changed vs
/// the store's previously established state — losing automatic offers and
/// pure reads answer `changed: false` with the actual accepted winner.
/// `native_sync` (Task 3) is the additive writeback status projection;
/// status-only updates (an unchanged record whose nativeSync moved) publish
/// with `changed: false` and fold on the client by `documentGeneration`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionNameUpdate {
    pub record: SessionNameRecord,
    pub document_generation: NameRevision,
    pub redirects: Vec<SessionNameRedirect>,
    pub changed: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub native_sync: Option<NativeSyncProjection>,
}

/// Rename request body (HTTP `PATCH /api/session-names` and the MCP-bridged
/// rename). Omitted `name_intent` defaults to `automatic`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RenameSessionNameRequest {
    pub target: SessionNameRef,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name_intent: Option<NameIntent>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub if_revision: Option<NameRevision>,
}

/// `session.name.updated` server→client broadcast payload. Published only
/// after a successful store commit/adoption, in local generation order; the
/// client folds by record and redirect revision, never arrival time.
/// `native_sync` mirrors the Task 3 status projection on the wire.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionNameUpdated {
    pub record: SessionNameRecord,
    pub document_generation: NameRevision,
    pub redirects: Vec<SessionNameRedirect>,
    pub changed: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub native_sync: Option<NativeSyncProjection>,
}

// ---------------------------------------------------------------------------
// Task 7 — legacy-name migration envelope
// ---------------------------------------------------------------------------

/// Where a legacy candidate's label lived. The scope breaks total-order ties:
/// `session > pane > source_tab > terminal > derived`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LegacyCandidateScope {
    Session,
    Pane,
    SourceTab,
    Terminal,
    Derived,
}

/// What protection the raw evidence claims for a legacy candidate.
/// `explicit_rename` is reliable explicit-rename evidence (only the old
/// durable session-rename ladder could produce it, and only server-side boot
/// evidence honors it); `legacy_flag` is an old manual/protected flag whose
/// human origin is UNRECOVERABLE (pane/tab user-set booleans, unsourced
/// title overrides); `none` is an unprotected label with a known automatic
/// origin.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LegacyProtectionEvidence {
    ExplicitRename,
    LegacyFlag,
    None,
}

/// A migration candidate's target: the canonical naming ref, or a legacy
/// identity that the server resolves through the existing canonical
/// identity/ledger — never a new key. `legacy_session` reuses the structured
/// provider/session restore identity (with the optional codex durability
/// evidence and original cwd); `legacy_terminal` names a terminal +
/// server instance for ledger resolution. The untagged representation
/// matches the TS `z.union`: the canonical ref carries its own `kind` tag
/// (`pending`/`session`) and the legacy variants carry `legacy_session`/
/// `legacy_terminal` through the flattened marker below. (The wire shape
/// is fixed by the plan's TS union — boxing the optional codex durability
/// evidence would buy nothing at this enum's call frequency.)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
#[allow(clippy::large_enum_variant)]
pub enum LegacyNameTarget {
    Canonical(SessionNameRef),
    #[serde(rename_all = "camelCase")]
    LegacySession {
        #[serde(flatten)]
        kind: LegacySessionKind,
        session_ref: SessionLocator,
        #[serde(skip_serializing_if = "Option::is_none")]
        codex_durability: Option<crate::common::CodexDurability>,
        #[serde(skip_serializing_if = "Option::is_none")]
        cwd: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    LegacyTerminal {
        #[serde(flatten)]
        kind: LegacyTerminalKind,
        terminal_id: String,
        server_instance_id: String,
    },
}

/// The `kind: "legacy_session"` discriminant (its own unit type so the
/// untagged variant above stays self-describing on the wire).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct LegacySessionKind {
    #[serde(rename = "kind")]
    pub kind: LegacySessionKindTag,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LegacySessionKindTag {
    LegacySession,
}

/// The `kind: "legacy_terminal"` discriminant.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct LegacyTerminalKind {
    #[serde(rename = "kind")]
    pub kind: LegacyTerminalKindTag,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LegacyTerminalKindTag {
    LegacyTerminal,
}

impl LegacyNameTarget {
    /// Convenience constructors matching the TS discriminated shapes.
    pub fn legacy_session(session_ref: SessionLocator) -> Self {
        Self::LegacySession {
            kind: LegacySessionKind {
                kind: LegacySessionKindTag::LegacySession,
            },
            session_ref,
            codex_durability: None,
            cwd: None,
        }
    }

    pub fn legacy_terminal(
        terminal_id: impl Into<String>,
        server_instance_id: impl Into<String>,
    ) -> Self {
        Self::LegacyTerminal {
            kind: LegacyTerminalKind {
                kind: LegacyTerminalKindTag::LegacyTerminal,
            },
            terminal_id: terminal_id.into(),
            server_instance_id: server_instance_id.into(),
        }
    }
}

/// The stable legacy-candidate id: the JSON encoding of the tuple
/// `[storageKey,deviceId,tabId,paneId,scope,provider,sessionId,name,source,
/// protectionEvidence,explicitRenameAt]` with absent entries `null`. Every
/// imported time must have evidence — never receipt/import time — so
/// `explicit_rename_at` is `None` unless the evidence carries a trustworthy
/// explicit-rename timestamp.
pub fn legacy_name_candidate_id(input: LegacyCandidateIdInput<'_>) -> String {
    let json = serde_json::json!([
        input.storage_key,
        input.device_id,
        input.tab_id,
        input.pane_id,
        input.scope,
        input.provider,
        input.session_id,
        input.name,
        input.source,
        input.protection_evidence,
        input.explicit_rename_at,
    ]);
    serde_json::to_string(&json).expect("the candidate id tuple serializes")
}

/// The identity fields of one legacy candidate for
/// [`legacy_name_candidate_id`].
#[derive(Debug, Clone, Copy)]
pub struct LegacyCandidateIdInput<'a> {
    pub storage_key: &'a str,
    pub device_id: Option<&'a str>,
    pub tab_id: Option<&'a str>,
    pub pane_id: Option<&'a str>,
    pub scope: LegacyCandidateScope,
    pub provider: Option<&'a str>,
    pub session_id: Option<&'a str>,
    pub name: &'a str,
    pub source: NameSource,
    pub protection_evidence: LegacyProtectionEvidence,
    pub explicit_rename_at: Option<i64>,
}

/// One legacy-name candidate. The stable candidate `id` is the JSON tuple
/// `[storageKey,deviceId,tabId,paneId,scope,provider,sessionId,name,source,
/// protectionEvidence,explicitRenameAt]` with absent entries `null`;
/// `explicit_rename_at` must carry evidence (never receipt/import time) and
/// is absent when no trustworthy explicit-rename time exists.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LegacyNameCandidate {
    pub id: String,
    pub target: LegacyNameTarget,
    pub name: String,
    pub source: NameSource,
    pub scope: LegacyCandidateScope,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub explicit_rename_at: Option<i64>,
    pub evidence_key: String,
    pub protection_evidence: LegacyProtectionEvidence,
}

/// One immutable raw evidence envelope: the bytes of one legacy storage
/// payload, preserved verbatim before any sanitization could clear it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LegacyEvidenceEnvelope {
    pub storage_key: String,
    pub raw: String,
}

/// The Task 7 import envelope. `version` is fixed at 1; `import_id` is the
/// persisted retry-stable id of one captured raw envelope (the server derives
/// its immutable per-import backup filename from it). Clients batch at most
/// 100 candidates per import.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LegacyNameImport {
    pub version: u32,
    pub import_id: String,
    pub evidence: Vec<LegacyEvidenceEnvelope>,
    pub candidates: Vec<LegacyNameCandidate>,
}

/// The import result: every acknowledged candidate id (per-candidate
/// acknowledgment, never an early whole-import shortcut) plus the updates for
/// the records the import touched (winners and unchanged winners alike).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LegacyImportResult {
    pub acknowledged: Vec<String>,
    pub names: Vec<SessionNameUpdate>,
}
